modules/storages/app/services/storages/nextcloud_managed_folder_sync_service.rb

```diff
@@ -246,33 +252,41 @@ def create_remote_folder(project_storage)
     def audit_last_project_folder(last_project_folder, project_folder_id)
       ApplicationRecord.transaction do
         success = last_project_folder.update(origin_folder_id: project_folder_id) &&
-                  last_project_folder.project_storage.update(project_folder_id:)
+          last_project_folder.project_storage.update(project_folder_id:)
 
         raise ActiveRecord::Rollback unless success
       end
     end
 
-    # @param group_folder [string] name of the Group Folder in Nextcloud.
-    # @param username [String] username for the integration user
-    # @param group [String] group that the user should be part of
+    # rubocop:disable Metrics/AbcSize
+    # @param root_folder_id [String] the id of the root folder
     # @return [ServiceResult]
-    def ensure_root_folder_permissions(group_folder, username, group)
-      info "Setting needed permissions for user #{username} and group #{group} on #{group_folder} group folder"
-
-      command_params = {
-        path: group_folder,
-        permissions: {
-          users: { username.to_sym => ALL_PERMISSIONS },
-          groups: { group.to_sym => PERMISSIONS_MAP[:read_files] }
-        }
-      }
-
-      set_permissions.call(storage: @storage, auth_strategy:, **command_params).on_failure do |service_result|
-        log_storage_error(service_result.errors, { folder: group_folder })
-        add_error(:ensure_root_folder_permissions, service_result.errors, options: { group:, username: }).fail!
-      end
+    def ensure_root_folder_permissions(root_folder_id)
+      username = @storage.username
+      group = @storage.group
+      info "Setting needed permissions for user #{username} and group #{group} on the root group folder."
+      permissions = [
+        { user_id: username, permissions: FILE_PERMISSIONS },
+        { group_id: group, permissions: [:read_files] }
+      ]
+
+      build_set_permissions_input_data(root_folder_id, permissions)
+        .either(
+          ->(input_data) do
+            set_permissions.call(storage: @storage, auth_strategy:, input_data:).on_failure do |service_result|
+              log_storage_error(service_result.errors, folder: "root", root_folder_id:)
+              add_error(:ensure_root_folder_permissions, service_result.errors, options: { group:, username: }).fail!
+            end
+          end,
+          ->(failure) do
+            log_validation_error(failure, root_folder_id:, permissions:)
+            ServiceResult.failure(result: failure.errors.to_h) # rubocop:disable Rails/DeprecatedActiveModelErrorsMethods
+          end
+        )
```

Comment:
Maybe we can make use of a `value_or` here so we don't have that many nested blocks.


```ruby

input_data = build_set_permissions_input_data(root_folder_id, permissions)
               .value_or do |failure|
               log_validation_error(failure, root_folder_id:, permissions:)
               return ServiceResult.failure(result: failure.errors.to_h)
             end 

set_permissions.call(storage: @storage, auth_strategy:, input_data:)
  .on_failure do |service_result|
     log_storage_error(service_result.errors, folder: "root", root_folder_id:)
     add_error(:ensure_root_folder_permissions, service_result.errors, options: { group:, username: }).fail!
  end
```